wrappers/python/openmm/app/xtcfile.py

```diff
@@ -0,0 +1,156 @@
+            if natoms != len(list(topology.atoms())):
+                raise ValueError(
+                    f"The number of atoms in the topology ({len(list(topology.atoms()))}) does not match the number of atoms in the XTC file ({natoms})"
+                )
+        else:
+            if os.path.isfile(self._filename) and os.path.getsize(self._filename) > 0:
+                raise FileExistsError(f"The file '{self._filename}' already exists.")
+
+    def _getNumFrames(self):
+        return get_xtc_nframes(self._filename.encode("utf-8"))
+
+    def writeModel(self, positions, unitCellDimensions=None, periodicBoxVectors=None):
+        """Write out a model to the XTC file.
+
+        The periodic box can be specified either by the unit cell dimensions
+        (for a rectangular box), or the full set of box vectors (for an
+        arbitrary triclinic box).  If neither is specified, the box vectors
+        specified in the Topology will be used. Regardless of the value
+        specified, no dimensions will be written if the Topology does not
+        represent a periodic system.
+
+        Parameters
+        ----------
+        positions : list
+            The list of atomic positions to write
+        unitCellDimensions : Vec3=None
+            The dimensions of the crystallographic unit cell.
+        periodicBoxVectors : tuple of Vec3=None
+            The vectors defining the periodic box.
+        """
+        if len(list(self._topology.atoms())) != len(positions):
+            raise ValueError("The number of positions must match the number of atoms")
+        if is_quantity(positions):
+            positions = positions.value_in_unit(nanometers)
+        if any(math.isnan(norm(pos)) for pos in positions):
+            raise ValueError(
+                "Particle position is NaN.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan"
+            )
+        if any(math.isinf(norm(pos)) for pos in positions):
+            raise ValueError(
+                "Particle position is infinite.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan"
+            )
+
+        self._modelCount += 1
+        if (
+            self._interval > 1
+            and self._firstStep + self._modelCount * self._interval > 1 << 31
+        ):
+            # This will exceed the range of a 32 bit integer.  To avoid crashing or producing a corrupt file,
+            # update the file to say the trajectory consisted of a smaller number of larger steps (so the
+            # total trajectory length remains correct).
+            self._firstStep //= self._interval
+            self._dt *= self._interval
+            self._interval = 1
+            with tempfile.NamedTemporaryFile() as temp:
+                nframes = self._getNumFrames()
+                coords, box, time, step = read_xtc(self._filename.encode("utf-8"))
```

Comment:
AFAIK XTC stores a constant number of atoms, but the rest of the parameters (box, step, time, coords and box) are stored per step. This means that this functionality requires going over every step.
There is this header comment in the xdrfile.h, which furthermore does not expose any functions to modify an existing file beyond appending frames:
```c++
/*
...
* We do not support getting or setting positions in XDR files, since it can
 * break in horrible ways for large (64-bit) files, resulting in silent data
 * corruption. Note that it works great to open/read/write 64-bit files if
 * your system supports it; it is just the random access we cannot trust!
```
I agree loading everything into memory is not a good idea, though. I implemented a function to read and write frame by frame C++ side, without loading the whole file into memory.
Still this requires storing the trajectory twice while this rewriting happens,  but I believe this is what it is.

Reply:
I mean storing the trajectory twice in the HDD, not in RAM.

Reply:
That's probably the best we can do then.  A trajectory that long could easily be up into the terabyte range.  Hopefully they'll have enough disk space to briefly store a duplicate copy.